sway/xdg_decoration.c

```diff
@@ -0,0 +1,73 @@
+#include <stdlib.h>
+#include "sway/desktop/transaction.h"
+#include "sway/server.h"
+#include "sway/tree/arrange.h"
+#include "sway/tree/view.h"
+#include "sway/xdg_decoration.h"
+#include "log.h"
+
+static void xdg_decoration_handle_destroy(struct wl_listener *listener,
+		void *data) {
+	struct sway_xdg_decoration *deco =
+		wl_container_of(listener, deco, destroy);
+	deco->view->xdg_decoration = NULL;
+	wl_list_remove(&deco->destroy.link);
+	wl_list_remove(&deco->surface_commit.link);
+	wl_list_remove(&deco->link);
+	free(deco);
+}
+
+static void xdg_decoration_handle_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct sway_xdg_decoration *decoration =
+		wl_container_of(listener, decoration, surface_commit);
+
+	bool csd = decoration->wlr_xdg_decoration->current_mode ==
+		WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_CLIENT_SIDE;
+	struct sway_view *view = decoration->view;
+
+	view_set_csd_from_client(view, csd);
```

Comment:
`current_mode` isn't the client preference, it's the current mode as chosen by the compositor.

To get the client preference, listen to `request_mode`.